File: Algorithm/C++/1000+/1017. Convert to Base -2.cc
/*
 * Created on Sun Apr 07 2019 20:25:25
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

 // bit manipulation
 
#include <string>


class Solution {
 public:
  std::string baseNeg2(int N) {
    if (N == 0) return "0";
    std::string result = "";
    int idx = 0, curr = 0, carry = 0;
    while (N > 0 || carry > 0) {
      int s = (N&1) + carry;
      curr = s % 2;
      carry = s / 2;
      result = std::to_string(curr) + result;
      if ((idx&1) > 0 && curr > 0) carry = 1;
      N >>= 1;
      idx++;
    }
    return result;
  }
};

File: Algorithm/Python/1000+/1017. Convert to Base -2.py
# -*- coding: utf-8 -*-
# Created on Wed Apr 03 2019 7:36:4
# Author: WuLC
# EMail: liangchaowu5@gmail.com

# bit manipulation
class Solution(object):
    def baseNeg2(self, N):
        """
        :type N: int
        :rtype: str
        """
        if N == 0:
            return "0"
        result = ""
        idx, carry = 0, 0
        while N or carry:
            s = (N&1) + carry
            curr, carry = s % 2, s / 2
            result = str(curr) + result
            if (idx&1) and curr:
                carry = 1
            N >>= 1
            idx += 1
        return result